utils/cmake/dreamcast.toolchain.cmake

```diff
@@ -0,0 +1,106 @@
+# CMake Toolchain file for targeting the Dreamcast or NAOMI with CMake.
+#  (c)2023 Falco Girgis, Luke Benstead
+#
+# This file is to be passed to CMake when compiling a regular CMake project
+# to cross-compile for the Dreamcast, using the KOS environment and settings.
+#
+#     cmake /path/to/src -DCMAKE_TOOLCHAIN_FILE=${KOS_BASE}/kos/utils/cmake/dreamcast.toolchain.cmake
+#
+#   or alternatively:
+#
+#     cmake /path/to/src -DCMAKE_TOOLCHAIN_FILE=${KOS_CMAKE_TOOLCHAIN}
+#
+# The toolchain enables the options required for the compiler to use the 
+# special SH4 instructions for sin/cos and invSqrt. Note that even when 
+# these options are present, they are not activated unless -ffast-math 
+# is also provied. Typically you will want to provide a project-level 
+# option for enabling this flag, which will make the toolchain's SH4 
+# instruction flags also take effect.
+#
+# Frame pointers are enabled in debug builds as these are required for 
+# stack traces and GDB. They are disabled in release.
+# 
+# The original toolchain file was created by Kazade for the Simulant 
+# engine who has graciously allowed the rest of the scene to warez it. 
+#
+# Special thanks to Matt Godbolt of Compiler Explorer, for supporting
+# the SH toolchains, which allowed us to figure out the precise magical
+# incantations of compiler flags for fast math.  
+
+##### Configure CMake System #####
+set(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_SYSTEM_VERSION 1)
+set(PLATFORM_DREAMCAST TRUE)
+
+##### Configure Cross-Compiler #####
+set(CMAKE_CROSSCOMPILING TRUE)
+set(CMAKE_C_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-gcc)
+set(CMAKE_CXX_COMPILER $ENV{KOS_CC_BASE}/bin/sh-elf-g++)
```

Comment:
This applies to a lot most of these set operations, but is there a reason the more specific environment variables aren't being used here (`KOS_CC` & `KOS_CCPLUS` for these two)? Maybe I've just been out of the loop for too long, but are there people with projects not using a typical `environ.sh` and instead using a much more minimal set of environment variables (at quick glance: `KOS_BASE`, `KOS_CC_BASE`, & optionally `KOS_SUBARCH` & `KOS_CMAKE_TOOLCHAIN`)?

Reply:
Oh, we have just been having a whole huge discussion about this exact thing in the SImulatnt discord server. I've been doing some testing by swapping to those two to see if we could unify them, which does have benefits in terms of maintenance and consistency, but a couple things to note here:

1) I'm trying to verify from at least 1 IDE (Qt Creator for me) and command-line. Qt Creator seems to get really angry with kos-cc/kos-c++ and claims its an invalid compiler that cannot generate a binary. It has something to do with the configuration step where it's calling it with a bunch of extra args. Gonna play with it more to figure it out.
2) This makes some opinionated decisions about optimization level, adding debug symbols, and turning on -WExtra, which others might not be cool with. HOWEVER, it does seem as though you can simply pass the opposite flags from CMake and have them properly override the defaults. I'm still trying to do a bunch of testing, because I'd hate to find out some default set here screws the project
3) General spookiness. None of the regular things in Cmake like CMAKE_C_FLAGS/CMAKE_CXX_FLAGS, etc etc will actually have values representative of what is getting passed to the compiler and may lead to some confusion/irritation. I'm no CMake purist, so I don't know that I care, but just putting that out there in case someone does.

Still looking into all of this right now.

Reply:
1. I don't think kos-cc/kos-c++ are strictly required since they're just shell script wrappers around sh-elf-gcc and adding the flags from the environment. In the case of the lines 37 & 38, the values `$ENV{KOS_CC_BASE}/bin/sh-elf-gcc` & `$ENV{KOS_CC}` are the same value in a standard kos toolchain & `environ.sh`.
2. My suggestion would be to go with the default flags from the base toolchain/environ/makefiles and add some of these more opinionated optimizations as optional options that can be turned on based on the projects' specific needs. Not sure if this works with toolchain included files though. If it does, this would look something like
```cmake
option(EXTRA_DEBUG "Build with extra debug options" off)

if(EXTRA_DEBUG)
    set(EXTRA_FLAGS ${EXTRA_FLAGS} -DFRAME_POINTERS -fno-omit-frame-pointer)
endif(EXTRA_DEBUG)
```
3. I think what I did for CrabEmu's CMake build, way back, was similar to what I suggested in 2. I built up flags in an extra variable, but instead of setting `CMAKE_C_FLAGS` or `CMAKE_CXX_FLAGS` directly, I used `add_definitions`. A quick look at the documentation looks like `add_compile_options` might be better, and optionally `add_compile_definitions` for `-D` flags. I'll admit that it's been awhile since I've been deep into CMake, so I can't really say if that's the best strategy here. 
4. This should be pretty easy to test manually with a small c file, then compile with `-fno-builtins` normally, then again with `-fmsrra -fmsca -ffast-math`, and finally with `-fno-builtins -fmsrra -fmsca -ffast-math`, optionally maybe one more with `-fno-builtins` last, and see if the objects are significantly different based on the order of the flags. In theory, GCC overrides based on the rightmost flag (and I don't have time right this second to verify that in the docs.) If it would be better to just not trust that the compiler potentially respects flag order for overrides, something that could be potentially done would be using a `foreach` to build up a temporary variable with the options while checking for, and skipping, `-fno-builtins`. Similar to 2, defaulting to the basic kos behavior, and using `option(FAST_MATH ` to trigger this behavior to skip `-fno-builtins` and switch in the other flags.

Reply:
Thank you for this, I'm about to have some free time available to get back to this PR, and you've given good feedback and direction for what to look into and test. I'll keep you posted!